content/departments/engineering/dev/tools/grpc.md

```diff
@@ -0,0 +1,76 @@
+# gRPC
+
+This document provides an overview and documentation for Sourcegraph-Specific gRPC usage, including information about gRPC services, monitoring, common issues, and additional resources. The purpose of this document is to assist developers in understanding and working with gRPC.
+
+## GRPC Monitoring Guide
+
+to view the gRPC metrics, go to the built-in Grafana dashboard. The goal of these metrics is to identify specific types of errors that we are aware of, which originate from the grpc/protobuf libraries we employ. Such errors often hint at potential configuration issues or bugs related to our grpc-specific application logic. By implementing these additions, we aim to gain better insights into these errors and investigate any necessary adjustments or fixes.
+
+example dashboard:
+
+[gitserver](https://sourcegraph.com/-/debug/grafana/d/gitserver/git-server?orgId=1) for example. Within that dashboard, there are two sections with panels for gRPC metrics called `GRPC server metrics` and `GRPC "interanl errors" metrics`.
+
+There is also a filter to only view metrics for a specific rpc method at the top of the datchboard. e.g. https://sourcegraph.com/-/debug/grafana/d/gitserver/git-server?orgId=1&var-alert_level=All&var-shard=All&var-method=Archive
+
+### Metrics
+
+#### GRPC server metrics
+
+- `Request rate per-method over 2m` - This is a good indicator of wether or not the rpc method is being used if at all.
+- `Error percentage per-method over 2m` - Captures server side errors that are not related to grpc itself. For example, if the rpc method is called and later given a cancellation signal, then it is expected that the rpc method will fail. In this case, the error percentage will be high, but it is not a cause for concern because the failures are unrelated to grpc itself.
```

Comment:
> ... Captures server side errors that are not related to grpc itself.

This is not quite right. The nuance here is that every gRPC operation returns a [status code](https://grpc.github.io/grpc/core/md_doc_statuscodes.html) (which is the same core idea as the [HTTP status codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status), just different). 

These charts plot responses with only "status.OK" against the total set of responses. Some core things to think about here are:

- **Both normal "application logic" errors and errors that _might_ be related to the specifics of the gRPC implementation are captured in this chart**. 
- There are many reasons for a method to fail that's unrelated to the exact transport mechanism used (this repository doesn't exist., etc.)
- Some failure modes are expected. For instance, [status.DeadlineExceeded / or status.Cancelled](https://grpc.github.io/grpc/core/md_doc_statuscodes.html) is frequently returned when the caller intentionally cancels the operation. However, this could be an entirely intentional part of our application logic (example: there is no need to continue streaming the output of a git archive call if the user cancelled the search or navigated away from the blob page). 
   	- See https://sourcegraph.com/-/debug/grafana/d/gitserver/git-server?viewPanel=100711&orgId=1&var-alert_level=All&var-shard=All&var-method=Archive and https://sourcegraph.com/-/debug/grafana/d/gitserver/git-server?viewPanel=100740&orgId=1&var-alert_level=All&var-shard=All&var-method=Archive
-